### DISCUSSION TITLE
Is it correct to think that the data resides in files and is only accessed as needed?

### DISCUSSION THREAD

**Discussion by EricGT** [EvgSkv/logica#19] [Q&A] [answered]
The question may seem odd at first but having worked with SWI-Prolog for some time, currently all of the data for use with Prolog queries needs to be loaded into memory to be used. I have a need that could eventually need access to more data than can be loaded into memory on a standard PC with 16GB of memory. Even going to 128GB may not be enough. Also I am not concerned with how long the queries take, so paying the price in time to access the data from hard drives is not an issue.

Does Logica work like a normal SQL database where the data can reside on hard drives and is only accessed as needed? Or does all of the data need to be loaded into memory before a query can be run? Or something else?

**Comment by EvgSkv** [comment, marked as answer]
Thanks for your question!

That's correct! Logica delegates execution of the program to a database engine, which means data can be stored on disks and is accessed as needed.

BigQuery, which is currently the only fully-supprted Logica's back-end, is Google Datawarehouse which is capable executing heavy queries over datasets of hundreds of terabytes. In this case the data is stored in a distributed file system. A reasonably complex logic program over a dataset of under 1 terabyte is likely to run under a minute.

If you want to try Logica on datasets of a few gigabytes then Google BigQuery free quota of 1TB would be sufficient for quite a few experimental runs.

A lot of functionality is already supported for PostgreSQL, which makes it possible to run queries inside a standard PC. In this case the data would be stored on the local disk and accessed as needed as well.

Please let me know if you have further questions.

**Reply by EricGT** [reply]
Thanks for the quick response!

> If you want to try Logica on datasets of a few gigabytes then Google BigQuery free quota of 1TB would be sufficient for quite a few experimental runs.

I see [BigQuery public datasets](https://cloud.google.com/bigquery/public-data) and [Google Cloud Free Program](https://cloud.google.com/free/docs/gcp-free-tier) which while nice have limits that convert into a pay plan; sorry can't do that at present. Will save for a latter day. 

However I do know enough about SWI-Prolog and PostgreSQL ([SWI-Prolog connecting to PostgreSQL via ODBC](https://swi-prolog.discourse.group/t/swi-prolog-connecting-to-postgresql-via-odbc/2404)) to possibly give this a trial run. 

FYI - While not exactly the same as Logica there is a related SWI-Prolog pack. See: [Constraint Query Language A high level interface to SQL databases ](https://www.swi-prolog.org/pldoc/doc_for?object=section(%27packages/cql.html%27))

Also in the light of using SWI-Prolog with SQL databases, the following is useful. [Wiki Discussion: Prolog in the mind set of SQL](https://swi-prolog.discourse.group/t/wiki-discussion-prolog-in-the-mind-set-of-sql/2661)

I wrote most of the Wiki article and Wiki discussion in looking for answers to my problem so Logica is something I do plan to understand in more detail, but I am not the creator of CQL.

Thanks.

**Reply by EvgSkv** [reply]
The decision to proceed with PostgreSQL makes sense.

You might be able to run queries on BigQuery without transition to pay plan. E.g. [here](https://support.google.com/googleapi/answer/6158867?hl=en) it is stated that you can use BigQuery on public datasets with billing disabled. And [COVID-19](https://pantheon.corp.google.com/marketplace/product/bigquery-public-datasets/covid19-public-data-program?project=active-grp&folder=&organizationId=) datasets are available to query for free until September 15th, 2021.

Having said that, your rationale for using PostgreSQL engine is very clear. An example of running a Logica program on PostgreSQL can be found [here](https://github.com/EvgSkv/logica/blob/main/examples/Logica_example_Avengers_and_PostgreSQL.ipynb). You can modify and run it interactively in [CoLab](https://colab.sandbox.google.com/github/EvgSkv/logica/blob/main/examples/Logica_example_Avengers_and_PostgreSQL.ipynb).
Let me know if you have any questions on this.

Thank you for the links about connection of SWI-Prolog with PostgreSQL, I'll study that.
